durable-storage/src/merkle_layer/tree.rs

```diff
@@ -46,3 +46,190 @@ impl Avl {
+        if balance_factor.abs() > 1 || node.balance_factor() != balance_factor {
+            (false, 0)
+        } else {
+            (true, std::cmp::max(left_height, right_height) + 1)
+        }
+    }
+
+    impl Avl {
+        fn iter(&self) -> AvlIterator {
+            match &self.root {
+                None => AvlIterator { state: vec![] },
+                Some(node) => {
+                    let mut state = vec![(node.clone(), AvlIterationState::Current)];
+                    let mut cnode = node.clone();
+                    while let Some(left_node) = cnode.left_ref() {
+                        state.push((left_node.clone(), AvlIterationState::Current));
+                        cnode = left_node.clone();
+                    }
+                    AvlIterator { state }
+                }
+            }
+        }
+
+        pub fn empty() -> Self {
+            Self { root: None }
+        }
+
+        pub fn height_and_balance_factor_sanity_check(&self) -> bool {
+            match &self.root {
+                None => true,
+                Some(node) => {
+                    let (ret, _) = height_and_balance_factor_sanity_check_helper(node.clone());
+                    ret
+                }
+            }
+        }
+    }
+
+    impl Iterator for AvlIterator {
+        type Item = Arc<MavlNode>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            while let Some((node, state)) = self.state.pop() {
+                match state {
+                    AvlIterationState::Left => {
+                        self.state.push((node.clone(), AvlIterationState::Current));
+                        match node.left_ref() {
+                            None => (),
+                            Some(left_node) => {
+                                self.state
+                                    .push((left_node.clone(), AvlIterationState::Left));
+                            }
+                        }
+                        continue;
+                    }
+                    AvlIterationState::Current => {
+                        self.state.push((node.clone(), AvlIterationState::Right));
+                    }
+                    AvlIterationState::Right => {
+                        match node.right_ref() {
+                            None => (),
+                            Some(right_node) => self
+                                .state
+                                .push((right_node.clone(), AvlIterationState::Left)),
+                        }
+                        continue;
+                    }
+                }
+                return Some(node);
+            }
+            None
+        }
+    }
+
+    fn compare_tree_to_reference(tree: &Avl, reference: &BTreeMap<Key, Bytes>) {
+        for (node, (key, value)) in tree.iter().zip(reference.iter()) {
+            assert_eq!(node.key(), key);
+            assert_eq!(node.data(), value);
+        }
```

Comment:
would it be worth asserting that there are also exactly the same number of KV pairs in both? I think possibly one edge case is not covered where the tree has (somehow) managed to drop part of the tail